elm-src/index.elm

```diff
@@ -0,0 +1,489 @@
+    , matchingOptions : NixOptions
+    , page : Int
+    , location : Navigation.Location
+    , selected : Maybe String
+    , status : PageStatus
+    , debounce_state : Debounce.State
+    }
+
+
+clamped : Model -> Model
+clamped model =
+    { model | page = (clamp 1 (total_pages model.matchingOptions) model.page) }
+
+
+update_options : Model -> NixOptions -> Model
+update_options model unsorted_options =
+    let
+        options =
+            List.sortBy option_sort_key unsorted_options
+
+        matchingOptions =
+            filtered options model.terms
+
+        page =
+            case model.selected of
+                Just option ->
+                    case which_page matchingOptions (opt_tuple_matches_name option) of
+                        Just p ->
+                            p
+
+                        Nothing ->
+                            model.page
+
+                Nothing ->
+                    model.page
+    in
+        clamped
+            ({ model
+                | options = options
+                , matchingOptions = matchingOptions
+                , page = page
+                , status = Loaded
+             }
+            )
+
+
+update_query : Model -> String -> Model
+update_query model query =
+    { model
+        | query = query
+        , terms = splitQuery query
+        , matchingOptions = filtered model.options model.terms
+        , page = 1
+        , selected = Nothing
+    }
+
+
+update_page : Model -> Int -> Model
+update_page model newPage =
+    clamped
+        { model
+            | page = newPage
+            , selected = Nothing
+        }
+
+
+select_option : Model -> String -> Model
+select_option model opt =
+    { model
+        | selected = Just opt
+    }
+
+
+deselect_option : Model -> Model
+deselect_option model =
+    { model
+        | selected = Nothing
+    }
+
+
+init : Navigation.Location -> ( Model, Cmd Msg )
+init location =
+    let
+        qs =
+            QueryString.parse location.search
+
+        query =
+            Maybe.withDefault "" (QueryString.one QueryString.string "query" qs)
+
+        page =
+            Maybe.withDefault 1 (QueryString.one QueryString.int "page" qs)
+
+        selected =
+            QueryString.one QueryString.string "selected" qs
+    in
+        ( Model query (splitQuery query) [] [] page location selected Loading Debounce.init
+        , getOptions
+        )
+
+
+splitQuery : String -> List String
+splitQuery query =
+    String.words (String.toLower query)
+
+
+
+-- UPDATE
+
+
+updateFromUrl : Navigation.Location -> Model -> Model
+updateFromUrl location model =
+    let
+        qs =
+            QueryString.parse location.search
+
+        query =
+            Maybe.withDefault "" (QueryString.one QueryString.string "query" qs)
+
+        page =
+            Maybe.withDefault 1 (QueryString.one QueryString.int "page" qs)
+
+        selected =
+            QueryString.one QueryString.string "selected" qs
+    in
```

Comment:
Seems like there's some shared code between `updateFromUrl` and `init`..

Reply:
Good catch. Thanks!